pkg/ipam/poolallocator/allocator.go

```diff
@@ -0,0 +1,254 @@
+
+	if err != nil {
+		return nil, ipallocator.MultiIPAllocator{}, err
+	}
+
+	allocators, err := a.initIPAllocators(ipPool)
+	if err != nil {
+		return nil, ipallocator.MultiIPAllocator{}, err
+	}
+	return ipPool, allocators, nil
+}
+
+func (a *IPPoolAllocator) appendPoolUsage(ipPool *v1alpha2.IPPool, ip net.IP, state v1alpha2.IPPoolUsageState, resource string) error {
+	usageEntry := v1alpha2.IPPoolUsage{
+		IPAddress: ip.String(),
+		State:     state,
+		Resource:  resource,
+	}
+
+	newPool := ipPool.DeepCopy()
+	newPool.Status.Usage = append(ipPool.Status.Usage, usageEntry)
+	_, err := a.crdClient.CrdV1alpha2().IPPools().UpdateStatus(context.TODO(), newPool, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Warningf("IP Pool %s update failed: %+v", ipPool.Name, err)
+		return err
+	}
+	klog.Infof("IP Pool update successful %s: %+v", ipPool.Name, newPool.Status)
+	return nil
+
+}
+
+func (a *IPPoolAllocator) removePoolUsage(ipPool *v1alpha2.IPPool, ip net.IP) error {
+
+	ipString := ip.String()
+	var newList []v1alpha2.IPPoolUsage
+	for _, entry := range ipPool.Status.Usage {
+		if entry.IPAddress != ipString {
+			newList = append(newList, entry)
+		}
+	}
+
+	if len(newList) == len(ipPool.Status.Usage) {
+		return fmt.Errorf("IP address %s was not allocated from IP pool %s", ip, ipPool.Name)
+	}
+
+	ipPool.Status.Usage = newList
+
+	_, err := a.crdClient.CrdV1alpha2().IPPools().UpdateStatus(context.TODO(), ipPool, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Warningf("IP Pool %s update failed: %+v", ipPool.Name, err)
+		return err
+	}
+	klog.Infof("IP Pool update successful %s: %+v", ipPool.Name, ipPool.Status)
+	return nil
+
+}
+
+// AllocateIP allocates the specified IP. It returns error if the IP is not in the range or already
+// allocated, or in case CRD failed to update its state.
+// In case of success, IP pool CRD status is updated with allocated IP/state/resource.
+// AllocateIP returns subnet detailes for the requested IP, as defined in IP pool spec.
+func (a *IPPoolAllocator) AllocateIP(ip net.IP, state v1alpha2.IPPoolUsageState, resource string) (v1alpha2.SubnetSpec, error) {
+	var subnetSpec v1alpha2.SubnetSpec
+	// Retry on CRD update conflict which is caused by multiple agents updating a pool at same time.
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		ipPool, allocators, err := a.readPoolAndInitIPAllocators()
+		if err != nil {
+			return err
+		}
+
+		index := len(allocators)
+		for i, allocator := range allocators {
+			if allocator.Has(ip) {
+				err := allocator.AllocateIP(ip)
+				if err != nil {
+					return err
+				}
+				index = i
+				break
+			}
+		}
+
+		if index == len(allocators) {
+			// Failed to find matching range
+			return fmt.Errorf("IP %v does not belong to IP pool %s", ip, a.ipPoolName)
+		}
+
+		err = a.appendPoolUsage(ipPool, ip, state, resource)
+
+		subnetSpec = *ipPool.Spec.IPRanges[index].SubnetSpec.DeepCopy()
```

Comment:
possibly a dumb question: is the DeepCopy call really needed here?

Reply:
Good point